### PR TITLE
Team sharp reg cache

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
TEAM SHARP:
1. Added environment config
2. Added registration cache. XCCL_TEAM_SHARP_RCACHE can be used to control registration cache 
3. Removed LL buffers

### CODE REVIEW

File: src/team_lib/sharp/xccl_sharp_collective.c [hunk 1]
```diff
@@ -3,27 +3,33 @@
...
+
+    status = ucs_rcache_get(ctx->rcache, address, length, PROT_READ|PROT_WRITE,
+                            NULL, &region);
+    if (status != UCS_OK) {
+        return XCCL_ERR_NO_MESSAGE;
     }
-    return XCCL_ERR_NO_RESOURCE;
+
+    *rregion = (xccl_sharp_rcache_region_t*)region;
+    return UCS_OK;
```
Comment (vspetrov) [inline src/team_lib/sharp/xccl_sharp_collective.c]: UCS_OK->XCCL_OK

File: src/team_lib/sharp/xccl_sharp_collective.c [hunk 2]
```diff
@@ -56,57 +61,60 @@ xccl_sharp_allreduce_init(xccl_coll_op_args_t* coll_args,
...
+            xccl_sharp_error("SHARP regmr failed for dst buffer\n");
         }
-        req->reduce_spec.sbuf_desc.buffer.ptr        = coll_args->buffer_info.src_buffer;
-        req->reduce_spec.sbuf_desc.buffer.mem_handle = send_mr;
-        req->reduce_spec.rbuf_desc.buffer.ptr        = coll_args->buffer_info.dst_buffer;
-        req->reduce_spec.rbuf_desc.buffer.mem_handle = recv_mr;
-        req->sharp_buf                               = NULL;
+    } else {
+        status = xccl_sharp_mem_rcache_reg(team_sharp_ctx,
+                                           coll_args->buffer_info.src_buffer,
```
Comment (vspetrov) [inline src/team_lib/sharp/xccl_sharp_collective.c]: Do we need to check here that src_buffer == dst_buffer (aka, INPLACE ) ? Then there is no need to register /deregister the same memory location twice.

File: src/team_lib/sharp/xccl_sharp_lib.c [hunk 3]
```diff
@@ -37,6 +46,27 @@ static inline void xccl_sharp_global_rand_state_init()
     xccl_sharp_rand_state_init(&xccl_sharp_global_rand_state);
 }
 
+static xccl_status_t xccl_sharp_lib_open(xccl_team_lib_h self,
+                                         xccl_team_lib_config_t *config) {
```
Comment (vspetrov) [inline src/team_lib/sharp/xccl_sharp_lib.c]: "{" to the new line

File: src/team_lib/sharp/xccl_sharp_lib.c [hunk 4]
```diff
@@ -113,13 +194,36 @@ xccl_sharp_create_context(xccl_team_lib_h lib, xccl_context_config_h config,
...
         free(ctx);
         return XCCL_ERR_NO_MESSAGE;
     }
+
+    ctx->rcache = NULL;
+    if (enable_rcache != UCS_NO) {
+        rcache_params.region_struct_size = sizeof(xccl_sharp_rcache_region_t);
+        rcache_params.alignment          = 64;
+        rcache_params.max_alignment      = (size_t)sysconf(_SC_PAGE_SIZE);
+        rcache_params.ucm_events         = XCCL_BIT(17) /*TODO: UCM_EVENT_VM_UNMAPPED */;
```
Comment (vspetrov) [inline src/team_lib/sharp/xccl_sharp_lib.c:225]: why XCCL_BIT?

**Review by vspetrov — APPROVED** [review]
@Sergei-Lebedev, you removed the small pre-registered eager buffers in favor of rcache - what is the perf implications? Usually, (i mean it is common) the small buffers are not registered - if the application will do many small allreduce calls with different buffers then it may lead to the very large increase of the number of rcache entries. Is there any eager buffers in libsharp_call ?

**Comment by Sergei-Lebedev** [comment]
@vspetrov, could you please check one more time?
small buffers in sharp team are returned back, added configuration option to control their size

**Review by vspetrov — APPROVED** [review]
👍

**Pull request merged**